### PR TITLE
Add test coverage for `HeaderElement` and `AcceptElement`

### PR DESCRIPTION
**Pull request by radez**
This is in preparation for issue #2014
Establishing tests to validate these objects so we can pull in the deprecated cgi.parse_header code

**What kind of change does this PR introduce?**
  - [x] tests/coverage improvement

### CODE REVIEW

File: cherrypy/test/test_httputil.py [hunk 1]
```diff
@@ -79,3 +80,82 @@ def test_invalid_status(status_code, error_msg):
     """Check that invalid status cause certain errors."""
     with pytest.raises(ValueError, match=error_msg):
         httputil.valid_status(status_code)
+
+
+@pytest.mark.parametrize(
+    'header_content,value,params',
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: I didn't add the corresponding linter here but a better style is having an iterable of strings ­— it reads better because it's sparse.
```suggestion
    ('header_content', 'value', 'params'),
```

File: cherrypy/test/test_httputil.py [hunk 2]
```diff
@@ -79,3 +80,82 @@ def test_invalid_status(status_code, error_msg):
...
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', {'q': '0.7'}),
+        ('text/plain',
+         'text/plain', {}),
+        ('text/xml',
+         'text/xml', {}),
+    ]
+)
+def test_header_element(header_content, value, params):
+    """ Test that value and params are parsed from headers passed to
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: FYI I'm pretty sure this will cause the linter you're fixing in #2028 to fail.

File: cherrypy/test/test_httputil.py (hunk 1 above)
Comment (radez) [reply cherrypy/test/test_httputil.py]: Fixed, I'll force push the update.

File: cherrypy/test/test_httputil.py (hunk 2 above)
Comment (radez) [reply cherrypy/test/test_httputil.py]: heh, yup it will. Fixed and included in the coming patch

**Comment by radez** [comment]
@webknjaz I pushed up some updates to address comments last week. Do I need to do anything else for this to go in?

**Comment by webknjaz** [comment]
Thanks for reminding! I'm going to take another look at this now as I'm on the PyCon US Sprints.

File: cherrypy/test/test_httputil.py [hunk 3]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+         'gzip', {'q': '0'}),
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', 0.8),
+    ]
+)
+def test_accept_element_raises_400(header_content, media_type, qvalue):
+    """ Test that headers passed to AcceptElement from httputils
+        raise an expected HTTPError """
+    httputil.AcceptElement.from_str(header_content)
+    assert pytest.raises(cherrypy.HTTPError)
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: This is an incorrect way of using `pytest.raises()`. There are two main problems with it:
1) It's always best to use the `match=` argument to state what the expectation is, otherwise, a too broad set of exception might make it pass.
2) This helper returns a context manager, and so it's meant to be used in the `with pytest.raises(Exc, match=r'^stuff$'):` block.

I suppose, it should look something like this:
```suggestion
    with pytest.raises(
            cherrypy.HTTPError,
            match=r'^Malformed HTTP header: `[^`]+`$',
    ):
        httputil.AcceptElement.from_str(header_content)
```

File: cherrypy/test/test_httputil.py [hunk 4]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+        ('gzip;q=1.0, identity; q=0.5, *;q=0',
+         'gzip', {'q': '0'}),
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', 0.8),
+    ]
+)
+def test_accept_element_raises_400(header_content, media_type, qvalue):
+    """ Test that headers passed to AcceptElement from httputils
+        raise an expected HTTPError """
+    httputil.AcceptElement.from_str(header_content)
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: Given the other comment I made, I'm surprised that this doesn't fail in the CI — it seems like you expect that it would.

File: cherrypy/test/test_httputil.py [hunk 5]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+def test_accept_element(header_content, media_type, qvalue):
+    """ Test that value and qvalue are parsed from headers passed to
+        AcceptElement from httputils """
+    acc_elem = httputil.AcceptElement.from_str(header_content)
+    assert acc_elem.value == media_type
+    assert acc_elem.qvalue == qvalue
+
+
+@pytest.mark.parametrize(
+    'header_content,media_type,qvalue',
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: I thought we had a linter for this... Let's enumerate the params list as an iterable of strings:
```suggestion
    ('header_content', 'media_type', 'qvalue'),
```

File: cherrypy/test/test_httputil.py [hunk 6]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+        Testing that the value and params are parsed from
+        headers passed to HeaderElement from httputils """
+    hdr_elem = httputil.HeaderElement.from_str(header_content)
+
+    assert hdr_elem.value == value
+    assert hdr_elem.params == params
+
+
+@pytest.mark.parametrize(
+    'header_content,media_type,qvalue',
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    ('header_content', 'media_type', 'qvalue'),
```
*(same)*

File: cherrypy/test/test_httputil.py [hunk 7]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+)
+def test_header_element(header_content, value, params):
+    """ Test that value and params are parsed from headers.
+
+        Testing that the value and params are parsed from
+        headers passed to HeaderElement from httputils """
+    hdr_elem = httputil.HeaderElement.from_str(header_content)
+
+    assert hdr_elem.value == value
+    assert hdr_elem.params == params
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: *(nit, nonblocking)* one assertion per test is usually preferred:
```suggestion
    assert (hdr_elem.value, hdr_elem.params) == (value, params)
```

File: cherrypy/test/test_httputil.py [hunk 8]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
     """Check that invalid status cause certain errors."""
     with pytest.raises(ValueError, match=error_msg):
         httputil.valid_status(status_code)
+
+
+@pytest.mark.parametrize(
+    ('header_content', 'value', 'params'),
+    [
+        ('application/x-www-form-urlencoded',
+         'application/x-www-form-urlencoded', {}),  # default_content_type
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: Here's an idea. The params get converted into long strings that are hard to follow, in the pytest output. To combat this, a typical solution would be to wrap each param via `pytest.param(..., id='...')` with that `id` made a human-readable kebab-case identifier for each case that would hint the reader what this parameter's high-level meaning is.
I got an example for you:
```suggestion
        pytest.param(
            'application/x-www-form-urlencoded',
            'application/x-www-form-urlencoded',
            {},
            id='default-content-type',
        ),
```

File: cherrypy/test/test_httputil.py [hunk 9]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+         'text/plain', {}),
+        ('text/xml',
+         'text/xml', {}),
+    ]
+)
+def test_header_element(header_content, value, params):
+    """ Test that value and params are parsed from headers.
+
+        Testing that the value and params are parsed from
+        headers passed to HeaderElement from httputils """
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]: Let's have this make more sense in the PEP 257 context:
```suggestion
    """Test that ``value`` and ``params`` are extracted from headers.

    This is a positive test case, checking that the value and
    params are parsed from headers that are being passed into
    the :py:class:`~cherrypy.httputil.HeaderElement` class.
    """
```

File: cherrypy/test/test_httputil.py [hunk 10]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+         'iso-8859-5, unicode-1-1', 0.8),
+        ('text/plain',
+         'text/plain', 1.0),
+        ('text/xml',
+         'text/xml', 1.0),
+    ]
+)
+def test_accept_element(header_content, media_type, qvalue):
+    """ Test that value and qvalue are parsed from headers passed to
+        AcceptElement from httputils """
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    """Test that ``value`` and ``qvalue`` are extracted from headers.

    This is being checked in the context of the
    :py:class:`~cherrypy.httputil.AcceptElement` class.
    """
```

**Review by webknjaz — COMMENTED** [review]
So.. I identified a few problems here. I'll try applying some patches and rebasing myself and will see where to go from there. Stay tuned!

File: cherrypy/test/test_httputil.py [hunk 11]
```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
...
+         'text/plain', {'q': '0.8, text/x-c'}),
+        ('gzip;q=1.0, identity; q=0.5, *;q=0',
+         'gzip', {'q': '0'}),
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', 0.8),
+    ]
+)
+def test_accept_element_raises_400(header_content, media_type, qvalue):
+    """ Test that headers passed to AcceptElement from httputils
+        raise an expected HTTPError """
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    """Check bad headers crash :class:`cherrypy.httputil.AcceptElement`.

    The expected exception is :exc:`~cherrypy.HTTPError`.
    """
```

File: cherrypy/test/test_httputil.py [hunk 12]
```diff
@@ -79,3 +80,91 @@ def test_invalid_status(status_code, error_msg):
...
+)
+def test_accept_element(header_content, media_type, qvalue):
+   """Test that ``value`` and ``qvalue`` are extracted from headers.
+
+   This is being checked in the context of the
+   :py:class:`~cherrypy.httputil.AcceptElement` class.
+   """
+    acc_elem = httputil.AcceptElement.from_str(header_content)
+    assert acc_elem.value == media_type
+    assert acc_elem.qvalue == qvalue
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    assert (acc_elem.value, acc_elem.qvalue) == (media_type, qvalue)
```

File: cherrypy/test/test_httputil.py [hunk 13]
```diff
@@ -79,3 +80,89 @@ def test_invalid_status(status_code, error_msg):
...
+)
+def test_header_element(header_content, value, params):
+    """Test that ``value`` and ``params`` are extracted from headers.
+
+    This is a positive test case, checking that the value and
+    params are parsed from headers that are being passed into
+    the :py:class:`~cherrypy.httputil.HeaderElement` class.
+    """
+    hdr_elem = httputil.HeaderElement.from_str(header_content)
+
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
```

File: cherrypy/test/test_httputil.py [hunk 14]
```diff
@@ -79,3 +80,88 @@ def test_invalid_status(status_code, error_msg):
     """Check that invalid status cause certain errors."""
     with pytest.raises(ValueError, match=error_msg):
         httputil.valid_status(status_code)
+
+
+@pytest.mark.parametrize(
+    ('header_content', 'value', 'params'),
+    [
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    (
```

File: cherrypy/test/test_httputil.py [hunk 15]
```diff
@@ -79,3 +80,88 @@ def test_invalid_status(status_code, error_msg):
...
+         'iso-8859-5, unicode-1-1', {'q': '0.8'}),
+        ('gzip;q=1.0, identity; q=0.5, *;q=0',
+         'gzip', {'q': '0'}),
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', {'q': '0.7'}),
+        ('text/plain',
+         'text/plain', {}),
+        ('text/xml',
+         'text/xml', {}),
+    ]
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    ),
```

File: cherrypy/test/test_httputil.py [hunk 16]
```diff
@@ -79,3 +80,88 @@ def test_invalid_status(status_code, error_msg):
...
+    params are parsed from headers that are being passed into
+    the :py:class:`~cherrypy.httputil.HeaderElement` class.
+    """
+    hdr_elem = httputil.HeaderElement.from_str(header_content)
+    assert (hdr_elem.value, hdr_elem.params) == (value, params)
+
+
+@pytest.mark.parametrize(
+    ('header_content', 'media_type', 'qvalue'),
+    [
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    (
```

File: cherrypy/test/test_httputil.py [hunk 17]
```diff
@@ -79,3 +80,88 @@ def test_invalid_status(status_code, error_msg):
...
+         'application/json', 1.0),
+        ('text/*, text/html, text/html;level=1, */*',
+         'text/*, text/html, text/html', 1.0),
+        ('iso-8859-5, unicode-1-1;q=0.8',
+         'iso-8859-5, unicode-1-1', 0.8),
+        ('text/plain',
+         'text/plain', 1.0),
+        ('text/xml',
+         'text/xml', 1.0),
+    ]
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    ),
```

File: cherrypy/test/test_httputil.py [hunk 18]
```diff
@@ -79,3 +80,88 @@ def test_invalid_status(status_code, error_msg):
...
+   This is being checked in the context of the
+   :py:class:`~cherrypy.httputil.AcceptElement` class.
+   """
+    acc_elem = httputil.AcceptElement.from_str(header_content)
+    assert (acc_elem.value, acc_elem.qvalue) == (media_type, qvalue)
+
+
+@pytest.mark.parametrize(
+    ('header_content', 'media_type', 'qvalue'),
+    [
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    (
```

File: cherrypy/test/test_httputil.py [hunk 19]
```diff
@@ -79,3 +80,88 @@ def test_invalid_status(status_code, error_msg):
...
+        ('audio/*; q=0.2, audio/basic',
+         'audio/*', {'q': '0.2, audio/basic'}),
+        ('text/plain; q=0.5, text/html, text/x-dvi; q=0.8, text/x-c',
+         'text/plain', {'q': '0.8, text/x-c'}),
+        ('gzip;q=1.0, identity; q=0.5, *;q=0',
+         'gzip', {'q': '0'}),
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', 0.8),
+    ]
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    ),
```

File: cherrypy/test/test_httputil.py [hunk 20]
```diff
@@ -79,3 +80,88 @@ def test_invalid_status(status_code, error_msg):
...
+    """
+    hdr_elem = httputil.HeaderElement.from_str(header_content)
+    assert (hdr_elem.value, hdr_elem.params) == (value, params)
+
+
+@pytest.mark.parametrize(
+    ('header_content', 'media_type', 'qvalue'),
+    (
+        ('application/x-www-form-urlencoded',
+         'application/x-www-form-urlencoded', 1.0),  # default_content_type
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
        pytest.param(
            'application/x-www-form-urlencoded',
            'application/x-www-form-urlencoded',
            1.0,
            id='default-content-type',
        ),
```

File: cherrypy/test/test_httputil.py [hunk 21]
```diff
@@ -79,3 +80,99 @@ def test_invalid_status(status_code, error_msg):
...
+        ('text/xml',
+         'text/xml', 1.0),
+    ),
+)
+def test_accept_element(header_content, media_type, qvalue):
+   """Test that ``value`` and ``qvalue`` are extracted from headers.
+
+   This is being checked in the context of the
+   :py:class:`~cherrypy.httputil.AcceptElement` class.
+   """
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py]:
```suggestion
    """Test that ``value`` and ``qvalue`` are extracted from headers.

    This is being checked in the context of the
    :py:class:`~cherrypy.httputil.AcceptElement` class.
    """
```

File: cherrypy/test/test_httputil.py [hunk 22]
```diff
@@ -79,3 +80,99 @@ def test_invalid_status(status_code, error_msg):
...
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', 0.8),
+    ),
+)
+def test_accept_element_raises_400(header_content, media_type, qvalue):
+    """Check bad headers crash :class:`cherrypy.httputil.AcceptElement`.
+
+    The expected exception is :exc:`~cherrypy.HTTPError`.
+    """
+    with pytest.raises(
```
Comment (webknjaz) [inline cherrypy/test/test_httputil.py:174]: @radez so as I implemented this, the problem with the expectation in the test got evident. Could you see if you can identify what inputs would actually lead to the validation error expected here...

**Comment by jaraco** [comment]
As mentioned in #2021, I'm re-targeting this PR for the `debt/pre-cgi-drop` branch while we work out the details so we can unblock that PR. The idea is - get tests passing against `debt/pre-cgi-drop` and `main` to uncover any regression introduced by dropping the cgi module.

**Review by webknjaz — CHANGES_REQUESTED** [review]
@radez so I rebased this on top of `main` and 4 tests don't pass — https://github.com/cherrypy/cherrypy/actions/runs/9754497486/job/26921518822?pr=2027#step:25:161, just as I predicted (https://github.com/cherrypy/cherrypy/pull/2027/files#r1608887187):
* `cherrypy/test/test_httputil.py::test_accept_element_raises_400[audio/*; q=0.2, audio/basic-audio/*-qvalue0]`
* `cherrypy/test/test_httputil.py::test_accept_element_raises_400[text/plain; q=0.5, text/html, text/x-dvi; q=0.8, text/x-c-text/plain-qvalue1]`
* `cherrypy/test/test_httputil.py::test_accept_element_raises_400[gzip;q=1.0, identity; q=0.5, *;q=0-gzip-qvalue2]`
* `cherrypy/test/test_httputil.py::test_accept_element_raises_400[da, en-gb;q=0.8, en;q=0.7-da, en-gb-0.8]`

Could you look into fixing them?
